### PR TITLE
Migrate tests to Kotlin, reorganise `java-tests`

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**
This PR migrates majority of tests into Kotlin and reorganises the project structure under `java-tests`.

## Changes to tests in details
 * Apart from conversion to Kotlin, tests were moved under the package `io.spine.test` with subpackages corresponding to a subsystem under the test.
 * The test suite classes got the `ITest` suffix because they are integration tests.
 * Custom assertions were consolidated under `io.spine.validation.assertions.Assertions.kt`.

## Build changes under `java-tests`
 * The modules were reorganised to consolidate majority of tests under the module `validating`.
 * The module `extra-definitions` was renamed to `consumer-dependency` to clarify the meaning.
 * Proto types to be validated and Java classes representing custom options and validating rules were moved under `testFixtures` to separate them from the test code.

### CODE REVIEW

File: java-tests/validating/src/testFixtures/kotlin/io/spine/validation/assertions/Assertions.kt [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+ * Assert the given `builder` produces an invalid message.
+ *
+ * @param builder The message builder.
+ * @return Violations received from building the message.
+ */
+@CanIgnoreReturnValue
+fun assertInvalid(builder: Message.Builder): List<ConstraintViolation> {
+    try {
+        val msg = builder.build()
+        return fail( "Expected an invalid message but got: ${msg.toJson()}")
```
Comment (armiol) [inline java-tests/validating/src/testFixtures/kotlin/io/spine/validation/assertions/Assertions.kt:65]: There is a redundant space symbol after `fail(`.

File: java-tests/runtime/src/test/kotlin/io/spine/validate/ValidateUntilitySpec.kt [hunk 2]
```diff
@@ -80,64 +71,4 @@ internal class ValidateUntilitySpec : UtilityClassTest<Validate>(Validate::class
 
         formatted shouldBe "test 1 test 2"
     }
-
-    @MuteLogging
-    @Nested
-    @DisplayName("test message changes upon `(set_once)` and")
-    internal inner class SetOnce {
```
Comment (yevhenii-nadtochii) [inline java-tests/runtime/src/test/kotlin/io/spine/validate/ValidateUntilitySpec.kt:87]: I thought we should preserve them until the deprecated members are removed.

File: java-tests/runtime/src/test/kotlin/io/spine/validate/option/FieldValidatingOptionSpec.kt [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            typeName = containingMessage.declaration().name().value()
+            msgFormat = errorMessage(context)
+        }
+        return value.nonDefault()
+            .filter { it.toString().length > maxLength }
+            .map { violation }
+            .collect(ImmutableList.toImmutableList())
+    }
+}
+@Immutable
```
Comment (yevhenii-nadtochii) [inline java-tests/runtime/src/test/kotlin/io/spine/validate/option/FieldValidatingOptionSpec.kt:148]: Please put a missing line.

File: java-tests/runtime/src/test/kotlin/io/spine/validate/option/PatternSpec.kt [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+    @Test
+    fun `validate with 'dot_all' modifier`() {
+        val message = AllThePatterns.newBuilder()
+            .setDotAll("ab" + System.lineSeparator() + "cd")
+            .buildPartial()
+        assertValid(message)
+    }
+
+    private fun patternStringFor(email: String): @NonValidated PatternStringFieldValue {
```
Comment (yevhenii-nadtochii) [inline java-tests/runtime/src/test/kotlin/io/spine/validate/option/PatternSpec.kt:129]: Can we have it on the file level?

File: java-tests/runtime/src/test/kotlin/io/spine/validate/option/RangeConstraintSpec.kt [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package io.spine.validate.option
```
Comment (yevhenii-nadtochii) [inline java-tests/runtime/src/test/kotlin/io/spine/validate/option/RangeConstraintSpec.kt:26]:
```suggestion
 */
 
package io.spine.validate.option
```

File: java-tests/validating/src/test/kotlin/io/spine/test/JavaMessageSmokeTest.kt [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    @Test
+    fun `create a valid message using 'build()'`() {
+        assertDoesNotThrow {
+            val number = valid.build()
+            check(number)
+        }
+    }
+
+    @Test
+    @DisplayName("throw `ValidationException` if the message is not valid")
```
Comment (yevhenii-nadtochii) [inline java-tests/validating/src/test/kotlin/io/spine/test/JavaMessageSmokeTest.kt:68]: A non-migrated display name.

File: java-tests/validating/src/test/kotlin/io/spine/test/options/AssumedRequiredITest.kt [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        @Test
+        fun `an id of which cannot be empty`() = assertValid(
+            CreateProject.newBuilder()
+                .setId(Identifier.newUuid())
+        )
+
+        @Test
+        fun `an id which cannot be a default message`() {
```
Comment (yevhenii-nadtochii) [inline java-tests/validating/src/test/kotlin/io/spine/test/options/AssumedRequiredITest.kt:63]:
```suggestion
        fun `an id of which cannot be a default message`() {
```

File: java-tests/validating/src/test/kotlin/io/spine/test/options/AssumedRequiredITest.kt [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    `an entity state` {
+
+        @Test
+        fun `which is an ID of the entity`() {
+            val msg = Project.newBuilder()
+            assertViolation(msg, "id")
+        }
+
+        @Test
+        @DisplayName("ID of which must be a non-empty value")
```
Comment (yevhenii-nadtochii) [inline java-tests/validating/src/test/kotlin/io/spine/test/options/AssumedRequiredITest.kt:99]: A non-migrated test name.

**Review by armiol — COMMENTED** [review]
@alexander-yevsyukov as discussed vocally, here are some of my comments that I have managed to write simultaneously with you addressing them.

File: java-tests/runtime/src/test/kotlin/io/spine/validate/option/FieldValidatingOptionSpec.kt [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+import org.junit.jupiter.api.Assertions.fail
+import org.junit.jupiter.api.DisplayName
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+
+@DisplayName("`FieldValidatingOption` should")
+internal class FieldValidatingOptionSpec {
+
+    @Test
+    fun `return empty value if option is not present in external or field constraints`() {
```
[the remark this answers is not in the document]
Comment (armiol) [reply java-tests/runtime/src/test/kotlin/io/spine/validate/option/FieldValidatingOptionSpec.kt:55]: As discussed vocally, we have NOT removed the external validation yet.

File: java-tests/runtime/src/test/kotlin/io/spine/validate/ValidateUntilitySpec.kt (hunk 2 above)
Comment (alexander-yevsyukov) [reply java-tests/runtime/src/test/kotlin/io/spine/validate/ValidateUntilitySpec.kt:87]: I hope we'd finish with Validation fixes and completion soon.

**Review by armiol — APPROVED** [review]

**Pull request merged**